### ISSUE TITLE
Unable to ensure data consistency when operation is canceled

### ISSUE THREAD

**Issue by wtct** [umbraco/Umbraco-CMS#13957]
### Which Umbraco version are you using? (Please write the *exact* version, example: 10.1.0)

11.2.0

### Bug summary

Hi Guys,

I have found that ensuring data consistency when operation is cancelled is not possible.

I explained it by detailed steps to reproduce.

It's caused by completing scope even if the operation is cancelled:

https://github.com/umbraco/Umbraco-CMS/blob/contrib/src/Umbraco.Core/Services/ContentService.cs#L1003

However, the cancellation when Saving & Publishing is working good:

https://github.com/umbraco/Umbraco-CMS/blob/contrib/src/Umbraco.Core/Services/ContentService.cs#L1152

Probably this issue can be solved only by implementation as below, but what if you have 3rd party solution which is canceling operation without a scope?

```
public class AnotherContentSavingNotificationHandler : INotificationHandler<ContentSavingNotification>
{
    private readonly IScopeProvider _scopeProvider;

    public AnotherContentSavingNotificationHandler(IScopeProvider scopeProvider)
    {
        _scopeProvider = scopeProvider;
    }

    public void Handle(ContentSavingNotification notification)
    {
        using (var _scope = _scopeProvider.CreateScope())
        {
            var entity = notification.SavedEntities.First();

            if (entity.HasIdentity)
                notification.Cancel = true;
            else
                _scope.Complete();
        }
    }
}
```

However, there is still an issue with changing name in the textbox field.

### Steps to reproduce

1. Let's imagine that you have situation as below:
```
public class TestComposer : IComposer
{
    public void Compose(IUmbracoBuilder builder)
    {
        builder.AddNotificationHandler<ContentSavingNotification, ContentSavingNotificationHandler>();
        builder.AddNotificationHandler<ContentSavingNotification, OtherContentSavingNotificationHandler>();
        builder.AddNotificationHandler<ContentSavingNotification, AnotherContentSavingNotificationHandler>();
    }
}

public class ContentSavingNotificationHandler : INotificationHandler<ContentSavingNotification>
{
    private readonly IContentService _contentService;

    public ContentSavingNotificationHandler(IContentService contentService)
    {
        _contentService = contentService;
    }

    public void Handle(ContentSavingNotification notification)
    {
        var entity = notification.SavedEntities.First();

        if (entity.HasIdentity)
        {
            var sibling = _contentService.Create("New sibling node", entity.ParentId, entity.ContentType.Alias);

            _contentService.Save(sibling);
        }
    }
}

public class OtherContentSavingNotificationHandler : INotificationHandler<ContentSavingNotification>
{
    public void Handle(ContentSavingNotification notification)
    {
        var entity = notification.SavedEntities.First();

        if (entity.HasIdentity)
            entity.Name = "Node with a new sibling node";
    }
}

public class AnotherContentSavingNotificationHandler : INotificationHandler<ContentSavingNotification>
{
    public void Handle(ContentSavingNotification notification)
    {
        var entity = notification.SavedEntities.First();

        if (entity.HasIdentity)
            notification.Cancel = true;
    }
}
```
2. Let's assume that **AnotherContentSavingNotificationHandler** is implemented in the 3rd party solution where the cancellation is implemented as described in the documentation: https://docs.umbraco.com/umbraco-cms/reference/notifications/contentservice-notifications#usage
3. Create a "Test node" and click the Save button.
4. Click the Save button once again in the "Test node".
5. Let's pay attention to the changed node name in te textbox field:
![image](https://user-images.githubusercontent.com/5636357/224780732-7c9e24ae-141c-42d2-8139-55879d6ca949.png)
7. Refresh the backoffice (reload node).
8. Let's take a look at the unchanged node name:
![image](https://user-images.githubusercontent.com/5636357/224781131-ece170cc-4fa1-4e72-a89a-e3883251a69d.png)
9. Let's pay attention to the "New sibling node" which was created.
10. Remove "New sibling node".
11. Open the "Test node".
12. Click Save & publish.
13. Let's take look at changed the node name in the textbox field:
![image](https://user-images.githubusercontent.com/5636357/224786428-24266f41-e528-4e23-ac77-b48edcccfa52.png)
14. Refresh backoffice (reload node).
15. Let's take a look at the unchanged node name.
16. Let's pay attention that the "New sibling node" was NOT created this time.

### Expected result / actual result

Expected result after canceled Save operation:

1. The visible name of the "Test node" shouldn't be changed to "Node with a new sibling node".
2. The "New sibling node" shouldn't be created.

**Comment by wtct** [comment]
Hi @Zeegaan,

Please take into account that there is a workaround but only if you have control over all cancelling operations. So if you have installed a 3rd party solutions which rely on cancellation then data could be corrupted.

cc: @piotrbach

**Comment by wtct** [comment]
> Strange indeed, I can definitely reproduce this 🕵️ As there is a workaround for now, I will mark this as up for grabs as we'd love some help with this 🙌

Hi @Zeegaan, I decided to prepare a new PR #16975 which should resolve the issue, because we would like to improve data consistency in our projects.

Please take into account that v13 is affected as well :)

It would be nice to include this PR together with: #16837. 

Thanks in advance!

cc: @nul800sebastiaan, @georgebid, @piotrbach

**Comment by ronaldbarendse** [comment]
It is actually a bug that the `Publish()` methos is missing a call to `scope.Complete()` (as there are multiple return statements):

https://github.com/umbraco/Umbraco-CMS/blob/e4dbf4ce9ff9ca70ed7c82c0c735a4d576aad692/src/Umbraco.Core/Services/ContentService.cs#L1207-L1236

**Comment by wtct** [comment]
> It is actually a bug that the `Publish()` method is missing a call to `scope.Complete()` (as there are multiple return statements):

To be honest with you @ronaldbarendse, for me it's a specific advantage we could rely on during ensuring data consistency, because in case of cancellation nothing will be persisted, especially changes applied in the notification handlers.

cc: @piotrbach

**Comment by ronaldbarendse** [comment]
> > It is actually a bug that the `Publish()` method is missing a call to `scope.Complete()` (as there are multiple return statements):
> 
> To be honest with you [@ronaldbarendse](https://github.com/ronaldbarendse), for me it's a specific advantage we could rely on during ensuring data consistency, because in case of cancellation nothing will be persisted, especially changes applied in the notification handlers.
> 
> cc: [@piotrbach](https://github.com/piotrbach)

Seems like you're trying to take advantage of something that isn't designed to work that way 😓 Cancelling a notification should only cancel the operation/entity it refers to, not cause the whole scope to rollback. Rollbacks should only occur when unexpected things happen (like an exception), since that could otherwise result in committing corrupted data to the database.

Although it might seem fine for scopes with a single operation, it becomes a problem if multiple (different) things are done in a single scope. To give a concrete example: Deploy content transfers are done in a single scope to ensure data consistency, so cancelling one `ContentSavingNotification` and not completing the scope would result in everything getting rolled back. This is definitely not the behavior we want.

If you still want this behavior, you can simply create a new (child) scope and not complete it in your own handler:

```c#
if (conditionToRollback)
{
    notification.Cancel = true;

    // Create a scope without completing it to cause a rollback
    using var scope = _scopeProvider.CreateScope();
}
```

**Comment by wtct** [comment]
> If you still want this behavior, you can simply create a new (child) scope and not complete it in your own handler:
> 
> if (conditionToRollback)
> {
>     notification.Cancel = true;
> 
>     // Create a scope without completing it to cause a rollback
>     using var scope = _scopeProvider.CreateScope();
> }

You are right @ronaldbarendse, but this is the same workaround I included during raising current issue.


> Probably this issue can be solved only by implementation as below, but what if you have 3rd party solution which is canceling operation without a scope?
> 
> ```
> public class AnotherContentSavingNotificationHandler : INotificationHandler<ContentSavingNotification>
> {
>     private readonly IScopeProvider _scopeProvider;
> 
>     public AnotherContentSavingNotificationHandler(IScopeProvider scopeProvider)
>     {
>         _scopeProvider = scopeProvider;
>     }
> 
>     public void Handle(ContentSavingNotification notification)
>     {
>         using (var _scope = _scopeProvider.CreateScope())
>         {
>             var entity = notification.SavedEntities.First();
> 
>             if (entity.HasIdentity)
>                 notification.Cancel = true;
>             else
>                 _scope.Complete();
>         }
>     }
> }
> ```

However, as I mentioned it's not covering all corner cases.

Let's imagine that you would like to store some data in a custom table, but this data has to be persisted together with the changes applied for a particular node. So you need to perform some custom operations in the notification handler for the "-ing" notification.

Unfortunately, once other notification handler will cancel the operation, only custom changes will be persisted, but the node will stay unchanged. By "other notification handler" I mean a handler over which you have no influence.

@piotrbach
